zkevm-circuits/src/evm_circuit/execution/chainid.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_WORD,
+        step::ExecutionState,
+        table::BlockContextFieldTag,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian};
+use halo2_proofs::{circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ChainIdGadget<F> {
+    same_context: SameContextGadget<F>,
+    chain_id: RandomLinearCombination<F, N_BYTES_WORD>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ChainIdGadget<F> {
+    const NAME: &'static str = "CHAINID";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CHAINID;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let chain_id = cb.query_rlc();
```

Comment:
you can do `let chain_id = cb.query_cell()` instead here because you never use the non-rlc form of chain in this gadget.

If you do that, you'll have to update the `assign_exec_step` to assign the RLC encoding of timestamp to the `timestamp` cell though.

Reply:
Sure, changed also. Thanks @z2trillion (also thanks to @ed255  to explain me some extra details to understand this issue).

What I do not understand here @z2trillion is why using RLC here will affect to the timestamp, also, could you elaborate a little please?

Reply:
I don't see how `timestamp` has anything to do here.

What is correct is the `query_cell` instead of `query_rlc` that's for sure. On that way you automatically get the word constructed! :)

Reply:
Sorry, I got confused about the block context field. Both timestamps in my original comment should have been chain_id

Reply:
I comment in the bottom to use RLC .